spec/declaration.dd

```diff
@@ -502,6 +502,44 @@ a = 2;         // sets `S.i` to `2`
 b = 4;         // sets `S.j` to `4`
 -----------
 
+
+$(H2 $(LNAME2 alias-reassignment, Alias Reassignment))
+
+        $(GRAMMAR
+        $(GNAME AliasReassignment):
+            $(GLINK_LEX Identifier) $(D =) $(GLINK StorageClasses)$(OPT) $(GLINK Type)
+            $(GLINK_LEX Identifier) $(D =) $(GLINK2 expression, FunctionLiteral)
+            $(GLINK_LEX Identifier) $(D =) $(GLINK StorageClasses)$(OPT) $(GLINK BasicType) $(GLINK2 function, Parameters) $(GLINK2 function, MemberFunctionAttributes)$(OPT)
+        )
+
+        $(P An alias declaration inside a template can be reassigned a new value.)
+
+        ---
+        template staticMap(alias F, T...)
+        {
+            alias A = AliasSeq!();
+            static foreach (t; T)
+                A = AliasSeq!(A, F!T); // alias reassignment
+            alias staticMap = A;
+        }
+        ---
+
+        $(P The $(I Identifier) must resolve to a lexically preceding $(GLINK AliasDeclaration).
+        Both must be members of the same $(GLINK2 template, TemplateDeclaration).
+        )
+
+        $(P The right hand side of the $(I AliasReassignment) replaces the right hand side of the
+        $(I AliasDeclaration).
+        )
+
+        $(P Once the $(I AliasDeclaration) has been referred to in any context other than the
+        right hand side of an $(I AliasReassignment) it can no longer be reassigned.
+        )
+
+        $(RATIONALE Alias reassignment can result in significantly faster compile times
+        and lowered memory consumption than the alternative recursive method.)
```

Comment:
Citation? I was under the impression that this is currently inconclusive.

Reply:
So was I.

Reply:
I don't need a test to know that:
```
malloc(5);
```
is faster and uses less memory than:
```
malloc(5);
malloc(5);
```
:-)

Reply:
https://github.com/dlang/phobos/pull/7756#issuecomment-764048209 Some plots on this. The performance increase is measurable but not particularly dramatic for `staticMap`

Reply:
> I don't need a test to know that:
> 
> ```
> malloc(5);
> ```
> 
> is faster and uses less memory than:
> 
> ```
> malloc(5);
> malloc(5);
> ```
> 
> :-)


But that's exactly what `static foreach` is doing, it will `syntaxCopy` the `foreach` body N-times.
The rationale should simply be to provide an imperative alternative for constructing alias sequences vs the recursive way.
The performance is just an implementation detail. I hope it gives more performance though.

Reply:
@WalterBright The consensus appears to be to revise this bit as:

> $(RATIONALE Alias reassignment can result in faster compile times and lowered memory consumption, 
and requires significantly simpler code than the alternative recursive method.)